Index: lib/Transforms/InstCombine/InstCombineSelect.cpp
===================================================================
--- lib/Transforms/InstCombine/InstCombineSelect.cpp    (revision 354644)
+++ lib/Transforms/InstCombine/InstCombineSelect.cpp    (working copy)
@@ -1652,7 +1652,7 @@
     Worklist.Add(Cond);
     return &SI;
   }
-
+#if 0
   if (SelType->isIntOrIntVectorTy(1) &&
       TrueVal->getType() == CondVal->getType()) {
     if (match(TrueVal, m_One())) {
@@ -1717,7 +1717,7 @@
       return new SExtInst(NotCond, SelType);
     }
   }
-
+#endif
   // See if we are selecting two values based on a comparison of the two values.
   if (FCmpInst *FCI = dyn_cast<FCmpInst>(CondVal)) {
     if (FCI->getOperand(0) == TrueVal && FCI->getOperand(1) == FalseVal) {