Index: lib/Transforms/InstCombine/InstructionCombining.cpp
===================================================================
--- lib/Transforms/InstCombine/InstructionCombining.cpp (revision 354644)
+++ lib/Transforms/InstCombine/InstructionCombining.cpp (working copy)
@@ -131,6 +131,10 @@
 MaxArraySize("instcombine-maxarray-size", cl::init(1024),
              cl::desc("Maximum array size considered when doing a combine"));

+static cl::opt<bool>
+DisableInstructionCombiner("disable-instruction-combiner", cl::init(false),
+                           cl::desc("Disable InstCombine Pass (default = off)"));
+
 // FIXME: Remove this flag when it is no longer necessary to convert
 // llvm.dbg.declare to avoid inaccurate debug info. Setting this to false
 // increases variable availability at the cost of accuracy. Variables that
@@ -3412,6 +3416,8 @@
     AssumptionCache &AC, TargetLibraryInfo &TLI, DominatorTree &DT,
     OptimizationRemarkEmitter &ORE, bool ExpensiveCombines = true,
     LoopInfo *LI = nullptr) {
+  if (DisableInstructionCombiner)
+    return false;
   auto &DL = F.getParent()->getDataLayout();
   ExpensiveCombines |= EnableExpensiveCombines;
